core/src/main/scala/BinaryTree.scala

```diff
@@ -0,0 +1,250 @@
+      case Branch(Some(a), l, r, _, _) => order.compare(x, a) match {
+        case LT => Branch(Some(a), l.remove(x), r)
+        case GT => Branch(Some(a), l, r.remove(x))
+        case EQ => (l, r) match {
+          case (BTNil(), BTNil()) => BinaryTree.empty
+          case (BTNil(), x) => x
+          case (x, BTNil()) => x
+          case (x, y) => {
+            val min = y.min
+
+            min match {
+              case Some(a) => Branch(min, x, y.remove(a))
+              case None() => x //<- this should never happen
+            }
+          }
+        }
+      }
+    }
+
+    balance(del(x))
+  }
+
+  def join(another: BinaryTree[A])(implicit order: Order[A]) = {
+
+    @tailrec def build(sub: BinaryTree[A], xs: List[A]): BinaryTree[A] = xs match {
+      case El()                 =>  sub
+      case Nel(h, t)            =>  build(sub + h, t)
+    }
+
+    another match {
+      case BTNil()              =>  this
+      case _                    =>  build(this, another.toLst())
+    }
+  }
+
+  def ++(another: BinaryTree[A])(implicit order: Order[A]) = join(another)
+
+  private def balance(tree: BinaryTree[A]): BinaryTree[A] = {
+    var t = tree.update
+
+    if (t.balanceFactor > 1) {
+      if (t.right.balanceFactor <= 0) {
+        t = Branch(t.value, t.left, rotate(t.right, 1), t.size, t.height)
+      }
+      t = rotate(t, -1)
+    } else {
+      if (t.balanceFactor < -1)
+      {
+        if (t.left.balanceFactor >= 0){
+          t = Branch(t.value, rotate(t.left, -1), t.right, t.size, t.height)
+        }
+        t = rotate(t, 1)
+      }
+    }
+
+    t
+  }
+
+  private def rotate(x: BinaryTree[A], direction: Int): BinaryTree[A] = {
+    if (direction < 0) {  // left
+
+      if (x == BTNil() || x.right == BTNil()) {
+        x
+      }
+      else {
+        val y = x.right
+        val a = Branch(x.value, x.left, BTNil(), x.size, x.height).update()
+
+        Branch(y.value, a, y.right, y.size, y.height).update()
+      }
+    }
+    else { //right
+      if (x == BTNil() || x.left == BTNil()) {
+        x
+      }
+      else {
+        val y = x.left
+        val a = Branch(x.value, BTNil(), x.right, x.size, x.height).update()
+
+        Branch(y.value, y.left, a, y.size, y.height).update()
+      }
+    }
+  }
+
+}
+
+
+
+case class Branch[A] private[dogs](value: Option[A],
+                     left: BinaryTree[A],
+                     right: BinaryTree[A],
+                     private [dogs] val size: Int = 0,
+                     private[dogs] val height: Int = 0) extends BinaryTree[A] {
+
+  private [dogs] override def balanceFactor: Int = this match {
+    case Branch(_, BTNil(), BTNil(), _, _)    =>  0
+    case Branch(_, BTNil(), r, _, _)          =>  r.height
+    case Branch(_, l, BTNil(), _, _)          =>  -1 * l.height
+    case Branch(_, l, r, _, _)                =>  r.height - l.height
+  }
+
+  private [dogs] override def update(): Branch[A] = this match {
```

Comment:
this method seems to only do one thing, which is to update the size and height members of a Branch method in case there is some situation where the size and height members were for some reason constructed with incorrect values?

Where is it possible that the size or height might possibly be incorrect that this might be anything but a no-op?